doc/fsck.reiser4.8
.\"						Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH fsck.reiser4 8 "05 February, 2004" reiser4progs "reiser4progs manual"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
fsck.reiser4 \- the program for checking and repairing reiser4 filesystem.
.SH SYNOPSIS
.B fsck.reiser4
[ options ] FILE
.SH DESCRIPTION
.B fsck.reiser4
is reiser4 filesystem check and repair program.
.SH CHECK OPTIONS
.TP
.B --check
the default action checks the consistency and reports, but does not repair any corruption that it finds. 
This option may be used on a read-only file system mount.
.TP
.B --fix
fixes minor corruptions that do not require rebuilding;
sets up correct values of bytes unsupported by kernel in the case of transparent compression.
.TP
.B --build-sb
fixes all severe corruptions in super blocks, rebuilds super blocks from scratch if needed.
.TP
.B --build-fs
fixes all severe fs corruptions, except super block ones; rebuilds reiser4 filesystem from the scratch if needed.
.TP
.B -L, --logfile
forces fsck to report any corruption it finds to the specified logfile rather then on stderr.
.TP
.B -n, --no-log
prevents fsck from reporting any kind of corruption.
.TP
.B -a, --auto
automatically checks the file system without any questions.
.TP
.B -q, --quiet
supresses gauges.
.TP
.B -r
ignored.
.SH PLUGIN OPTIONS
.TP
.B --print-profile
prints the plugin profile. This is the set of default plugins used for all
parts of a filesystem -- format, nodes, files, directories, hashes, etc. If
--override is specified, then prints modified plugins.
.TP
.B -l, --print-plugins
prints all plugins libreiser4 know about.
.TP
.B -o, --override TYPE=PLUGIN, ...
overrides the default plugin of the type "TYPE" by the plugin "PLUGIN"
in the plugin profile.
.SH COMMON OPTIONS
.TP
.B -V, --version
prints program version
.TP
.B -?, -h, --help
prints program help
.TP
.B -q, --quiet
suppress messages.
.TP
.B -y, --yes
assumes an answer 'yes' to all questions.
.TP
.B -f, --force
forces fsck to use whole disk, not block device or mounted partition.
.TP
.B -p, --preen
automatically repair minor corruptions on the filesystem.
.TP
.B -c, --cache N
tunes number of nodes in the libreiser4 tree buffer cache
.RS
.SH REPORTING BUGS
Report bugs to <reiserfs-devel@vger.kernel.org>
.SH SEE ALSO
.BR debugfs.reiser4(8),
.BR mkfs.reiser4(8),
.BR measurefs.reiser4(8)
.SH AUTHOR
This manual page was written by Vitaly Fertman <vitaly@namesys.com>


ChangeLog
This file is part of reiser4progs package

Copyright (C) 2001, 2002, 2003, 2004 by Hans Reiser, licensing governed by
reiser4progs/COPYING.

This file may be modified and/or distributed without restriction. This is
not an invitation to misrepresent the history of reiser4progs.

2015-08-14
	1.1.0 version
2015-08-14
	Format release 4.0.1;
	Support checksum for formatted nodes;
	NODE41 format with 32-bit field for checksums;
2014-06-28
	1.0.9 version
2014-06-28
	discard option to mkfs -- discard device at filesystem creation
	time (for SSD drives)

	Fixup in librepair/twig_scan.c: cb_item_region_check().
2013-05-03
	1.0.8 version
2013-05-03
	Fixups in fsck.reiser4: repair a special kind of corruptions
	appeared due to software bugs in earlier kernels.

	mkfs.reiser4 now enables intelligent compression with lzo1
	algorithm by default. If you want to create reiser4 without
	compression, then override the default plugin set with the
	"-o create=reg40" option.
2008-01-09
	1.0.7 version.
2008-01-09
	Alias -p "preen" to -a "auto" for fsck.
	Bugfixes:
	- fixed bugs in CryptoCompression object plugin (incorrect handling
	of corrupted compressed bodies).
	- suppress messages about wrong bytes for CryptoCompression
	objects	as in most cases it is because kernel doesn't support
	i_blocks and i_bytes for such objects because of performance
	issues. If wrong bytes is the result of fatal corruption, then
	fsck will detect and report it.
2006-02-22
	1.0.6 version.
2006-01-13
	Version control data is added into format40 disk format plugin.
	Version control data is added into backup blocks.
	Recover filesystems of both old and new formats - temporary.
	UPDATE_BACKUP_IS_NEEDED is added -- handle correctly cases when 
	disk format is updated and backup blocks are not; fsck correctly.
2005-12-09
	Heir plugin SET (HSET) is added.
2005-11-21
	Bugfixes:
	- a lot of fixes in compression tail item plugin;
	- a lot of fixes in CryptoCompression object plugin;
	- a lot of fixes in common object plugin repair code;
	- in spec file.
2005-11-08
	A lot of speedups.
	A lot of bugfixes.
	Demos/busy program update: crypto compression test is added.
2005-10-31
	Bugfixes in truncate, clubber and check_struct methods of all object
	plugins.
2005-10-28
	Compression plugin is added. 
	CryptoCompression plugin is adjusted to handle compressed files.
2005-10-14
	Bugfixes:
	- check the path given on a symlink creation;
	- check the possible file position overflow;
	- handle the cluster size of ctail items correctly.
2005-09-31
	Crypto-Compression Object plugins is added.
	Fix the object creation: inherit always from the parent.
	Fix the pset storage policy.
2005-08-31
	Compression tail item (ctail) plugin is added.
2005-08-16
	The full support for Crypto-Compression plugins is added into PluginSet
	and libreiser4 profile.
	Crypto SD extention is edded.
2005-08-12
	1.0.5 version.
2005-08-10
	Installation fix.
	Return value fixes.
	Backup block handling fixes.
2005-06-28
	Ability to link against libaal placed at arbitrary location is fixed.
	Unlink dir fix.
	Stat40 check_struct fix.
	Man pages for not ready utils are removed.
	Handle new (crc related) fields in plugin set correctly.
	Flush fs metadata fix (interrupt related).
	Do not flush dirty nodes if fsck fails.
	Recalculate bitmap checksums at the end of fsck if they are wrong.
	Do not update bitmaps after every fsck pass if no metadata found.
2005-03-23
	nomkid option to fsck -- ability to recover reiser4 w/out mkid check.
2005-03-21
	Large file handling fixes.
2005-03-16
	Some casting fixes.
2005-02-21
	Reg40 file body convertion fixes.
2005-02-20
	The 1.0.4 version.
2005-02-19
	A bugfix in file body convertion code.
2005-02-17
	Enable libminimal by default.
	Do not configure empty utils(resizer, cpfs).
2005-01-30
	Some bug fixes in tree balancing, fs check, syncing code.
2005-01-25
	Some bug fixes in tree balancing code.
2005-01-20
	Bug fixes:
	- in backup layout traverse code;
	- in insert extent item-by-item code;
2005-01-17
	A way to recover fs metadata independently of the mkfs id is added.
	Bug fixes:
	- in repair code then a node has a pointer to itself;
	- in not aligned access to different object on-disk data.
2004-12-23
	Bug fixes:
	- in initializing object id fields in SB;
	- in free blocks accounting;
	- in journal unpack code.
2004-12-12
	Some bugfixes:
	- in object/extent truncate code are fixed.
	- in backup handling code is fixed.
	- sware about libaal version correctly.
2004-12-04
	The version is 1.0.3
2004-12-01
	The Pack/unpack journal code is written.
	Cleanups in program options.
2004-11-25
	Build either all or none progs statically.
	The recover from unused space code is added. Could be used as undelete.
2004-11-24
	The backup code is written.
2004-10-29
	A bunch of bugs is fixed.
2004-10-21
	Object detach/unlink methods are fixed.
	Fix the handling of special entries ('.', '..') in directory objects.
	The version is 1.0.2.
2004-10-15
	File body write code (namely extent40 write) is rewritten
	and does work now.
2004-10-12
	Make allocator CRC endian safe
2004-10-07
	A bunch of bugs is fixed in the balancing code
2004-09-22
	Change "stand alone" to "minimal footprint"
2004-09-21
	Get rid of key collision problem is tree_scan
2004-09-16
	sd_bytes fix for files larger then 4G.
2004-09-14
	Make ldconfig to run automatically on 'make install'.
2004-09-13
	Make r5 hash working correctly with extended ASCII characters.
2004-08-24
	Some bugs are fixed:
	- in tree_scan method;
	- in reading extent items;
	- in man pages.
	Some optimisations for stand-alone are added.
	Version is 1.0.1.
2004-08-09
	1.0.0 version is released.


progs/fsck/fsck.c
/* Copyright (C) 2001-2005 by Hans Reiser, licensing governed by
   reiser4progs/COPYING.
   
   fsck.c -- reiser4 filesystem checking and recovering program. */

#ifndef _GNU_SOURCE
#  define _GNU_SOURCE
#endif

#include <fsck.h>

#define fsck_opt(parse, flag) aal_test_bit(&(parse)->options, flag)

static void fsck_print_usage(char *name) {
	fprintf(stderr, "Usage: %s [ options ] FILE\n", name);

	fprintf(stderr, 
		"Fsck options:\n"
		"  --check                       checks the consistency (default)\n"
		"  --fix                         fixes minor corruptions\n"
		"  --build-sb                    rebuilds the super block\n"
		"  --build-fs                    rebuilds the filesystem\n"
		"\n"
		"  -L, --logfile file            complains into the file\n"
		"  -n, --no-log                  makes fsck to not complain\n"
		"  -a, --auto                    automatically checks the consistency\n"
		"                                without any questions.\n"
		"  -q, --quiet                   supresses gauges\n"
		"  -r                            ignored\n"
		"Plugins options:\n"
		"  --print-profile               prints the plugin profile.\n"
		"  -l, --print-plugins           prints all known plugins.\n"
		"  -o, --override TYPE=PLUGIN    overrides the default plugin of the type\n"
	        "                                \"TYPE\" by the plugin \"PLUGIN\" in the\n"
		"                                profile.\n"
		"Common options:\n"
		"  -?, -h, --help                prints program usage.\n"
		"  -V, --version                 prints current version.\n"
		"  -y, --yes                     assumes an answer 'yes' to all questions.\n"
		"  -f, --force                   makes fsck to use whole disk, not block\n"
		"                                device or mounted partition.\n"
		"  -p, --preen                   automatically repair the filesysem.\n"
		"  -c, --cache N                 number of nodes in tree buffer cache\n");
}

#define WARNING \
"*******************************************************************\n"\
"This is an EXPERIMENTAL version of fsck.reiser4. Read README first.\n"\
"*******************************************************************\n"

static errno_t fsck_ask_confirmation(fsck_parse_t *data, char *host_name) {
	fprintf(stderr, WARNING "\n");
	aal_mess("Fscking the %s block device.", host_name);
	
	switch (data->sb_mode) {
	case RM_CHECK:
		aal_mess("Will check the consistency of "
			 "the Reiser4 SuperBlock.");
		break;
	case RM_FIX:
		aal_mess("Will fix minor corruptions of "
			 "the Reiser4 SuperBlock.");
		break;
	case RM_BUILD:
		aal_mess("Will build the Reiser4 SuperBlock.");
		break;
	default:
		break;
	}
	
	switch (data->fs_mode) {
	case RM_CHECK:
		aal_mess("Will check the consistency of "
			 "the Reiser4 FileSystem.");
		break;
	case RM_FIX:
		aal_mess("Will fix minor corruptions of "
			 "the Reiser4 FileSystem.");
		break;
	case RM_BUILD:
		aal_mess("Will build the Reiser4 FileSystem.");
		break;
	case RM_BACK:
		aal_mess("Will rollback changes saved in '%s' back "
			"onto (%s).", data->backup_file, host_name);
		break;
	default:
		break;
	}
	
	if (!fsck_opt(data, FSCK_OPT_YES)) {
		if (aal_yesno("Continue?") == EXCEPTION_OPT_NO) 
			return USER_ERROR;
	}
     
	return NO_ERROR; 
}

static void fsck_init_streams() {
	int ex;

	for (ex = 0; ex < EXCEPTION_TYPE_LAST; ex++)
		misc_exception_set_stream(ex, stderr);
	
	misc_exception_set_stream(EXCEPTION_TYPE_INFO, stdout);
}

enum fsck_mode {
	RM_SHOW_PARM = RM_LAST,
	RM_SHOW_PLUG = RM_LAST + 1
};

static errno_t fsck_init(fsck_parse_t *data, 
			 struct aal_device_ops *ops, 
			 int argc, char *argv[]) 
{
	static int mode = RM_CHECK, sb_mode = 0, fs_mode = 0;
	FILE *stream = NULL;
	char override[4096];
	int option_index;
	errno_t ret = 0;
	uint32_t cache;
	int mounted, c;

	static struct option options[] = {
		/* FSCK modes. */
		{"check", no_argument, &mode, RM_CHECK},
		{"fix", no_argument, &mode, RM_FIX},
		{"build-sb", no_argument, &sb_mode, RM_BUILD},
		{"build-fs", no_argument, &fs_mode, RM_BUILD},
		{"print-profile", no_argument, &mode, RM_SHOW_PARM},
		{"print-plugins", no_argument, &mode, RM_SHOW_PLUG},
		/* Fsck hidden modes. */
		{"rollback", no_argument, &mode, RM_BACK},
		/* Fsck options */
		{"logfile", required_argument, 0, 'L'},
		{"version", no_argument, NULL, 'V'},
		{"help", no_argument, NULL, 'h'},
		{"quiet", no_argument, NULL, 'q'},
		{"yes", no_argument, NULL, 'y'},
		{"no-log", no_argument, NULL, 'n'},
		{"auto", no_argument, NULL, 'a'},
		{"force", no_argument, NULL, 'f'},
		{"preen", no_argument, NULL, 'p'},
		{"cache", required_argument, 0, 'c'},
		{"override", required_argument, NULL, 'o'},
		/* Fsck hidden options. */
		{"passes-dump", required_argument, 0, 'U'},
		{"backup", required_argument, 0, 'b'},
		{"bitmap", required_argument, 0, 'B'},
		{"unused", no_argument, NULL, 'u'},
		{"oldfs", no_argument, NULL, 'O'},
		{"nomkid", no_argument, NULL, 'N'},
		{0, 0, 0, 0}
	};

	fsck_init_streams();
	aux_gauge_set_handler(misc_progress_handler, GT_PROGRESS);
	memset(override, 0, sizeof(override));
	data->logfile = stderr;

	if (argc < 2) {
		fsck_print_usage(argv[0]);
		return USER_ERROR;
	}

	while ((c = getopt_long(argc, argv, "L:VhnqafU:b:r?dB:plo:c:uyON", 
				options, &option_index)) >= 0) 
	{
		switch (c) {
		case 0:
			/* Long options. */
			break;
		case 'L':
			if ((stream = fopen(optarg, "w")) == NULL) {
				aal_fatal("Cannot not open the "
					  "logfile (%s).", optarg);
				goto oper_error;
			} 
			
			data->logfile = stream;		
			break;
		case 'n':
			data->logfile = NULL;
			break;
		case 'U':
			break;
		case 'b':
			data->backup_file = optarg;
			
			break;
		case 'f':
			aal_set_bit(&data->options, FSCK_OPT_FORCE);
			break;
		case 'p':
			/* Fall through to auto, as preen is an alias for -a */
		case 'a':
			aal_set_bit(&data->options, FSCK_OPT_AUTO);
			break;
		case 'h': 
		case '?':
			fsck_print_usage(argv[0]);
			goto user_error;
		case 'V': 
			misc_print_banner(argv[0]);
			goto user_error;
		case 'q':
			aux_gauge_set_handler(NULL, GT_PROGRESS);
			break;
		case 'y':
			aal_set_bit(&data->options, FSCK_OPT_YES);
			break;
		case 'r':
			break;
		case 'd':
			aal_set_bit(&data->options, FSCK_OPT_DEBUG);
			break;
		case 'B':
			data->bitmap_file = optarg;
			break;
		case 'c':
			if ((cache = misc_str2long(optarg, 10)) == INVAL_DIG) {
				aal_fatal("Invalid cache value specified (%s).",
					  optarg);
				return USER_ERROR;
			}

			misc_mpressure_setup(cache);
			break;
		case 'l':
			mode = RM_SHOW_PLUG;
			break;
		case 'o':
			aal_strncat(override, optarg, aal_strlen(optarg));
			aal_strncat(override, ",", 1);
			break;
		case 'u':
			aal_set_bit(&data->options, FSCK_OPT_WHOLE);
			break;
		case 'O':
			aal_set_bit(&data->options, FSCK_OPT_OLD);
			break;
		case 'N':
			aal_set_bit(&data->options, FSCK_OPT_NOMKID);
			break;
		}
	}
	
	/* Initializing libreiser4 with factory sanity check */
	if (libreiser4_init()) {
		aal_fatal("Cannot initialize the libreiser4.");
		goto oper_error;
	}

	if (aal_strlen(override) > 0) {
		override[aal_strlen(override) - 1] = '\0';

		aal_mess("Overriding the plugin profile by \"%s\".", 
			 override);
		
		if (misc_profile_override(override))
			goto oper_error;
	}

	if (mode == RM_SHOW_PARM) {
		misc_profile_print();
		goto user_error;
	}

	if (mode == RM_SHOW_PLUG) {
		misc_plugins_print();
		goto user_error;
	}

	if (optind != argc - 1) {
		fsck_print_usage(argv[0]);
		goto user_error;
	}
	
	if (aal_test_bit(&data->options, FSCK_OPT_AUTO)) {
		aal_mess("%s %s", argv[0], argv[optind]);
		exit(NO_ERROR);
	}

	data->sb_mode = sb_mode ? sb_mode : mode;
	data->fs_mode = fs_mode ? fs_mode : mode;
  
	if (data->backup_file) {
		data->backup = fopen(data->backup_file, 
				     mode == RM_BACK ? 
				     "r" : "w+");
		
		if (!data->backup) {
			aal_fatal("Cannot not open the backup file (%s).", 
				  optarg);
			goto oper_error;
		}
	}
	
	/* Check if device is mounted and we are able to fsck it. */ 
	if ((mounted = misc_dev_mounted(argv[optind])) > 0) {
		if (mounted == MF_RW) {
			aal_fatal("The partition (%s) is mounted with write "
				  "permissions, cannot fsck it.", argv[optind]);
			goto user_error;
		}
		
		aal_set_bit(&data->options, FSCK_OPT_RO);

		if (data->sb_mode != RM_CHECK || data->fs_mode != RM_CHECK) {
			/* If not CHECK mode, lock the process in the memory. */
			if (mlockall(MCL_CURRENT)) {
				aal_fatal("Failed to lock the process in the "
					  "memory: %s.", strerror(errno));
				goto oper_error;
			}
		}
	}
    
	if (!(data->host_device = aal_device_open(ops, argv[optind], 
						  512, O_RDONLY))) 
	{
		aal_fatal("Cannot open the partition (%s): %s.",
			  argv[optind], strerror(errno));
		goto oper_error;
	}

	misc_exception_set_stream(EXCEPTION_TYPE_FSCK, data->logfile);
	return fsck_ask_confirmation(data, argv[optind]);
	
 user_error:
	ret = USER_ERROR;
	goto error;
 oper_error:
	ret = OPER_ERROR;
 error:
	if (data->logfile && data->logfile != stderr)
		fclose(data->logfile);

	if (data->backup)
		fclose(data->backup);
	
	return ret;
}

static void fsck_fini(fsck_parse_t *data) {
	if (data->logfile != NULL && data->logfile != stderr)
		fclose(data->logfile);

	backup_fini();
}

static void fsck_time(char *string) {
	time_t t;

	time(&t);
	fprintf(stderr, "***** %s %s", string, ctime (&t));
}

/* Open the fs and init the tree. */
static errno_t fsck_check_init(repair_data_t *repair, 
			       aal_device_t *host, 
			       FILE *backup,
			       uint8_t sb_mode,
			       uint8_t fs_mode) 
{
	aal_stream_t stream;
	int flags = 0;
	int reop = 0;
	count_t len;
	errno_t res;
	
	/* Reopen device RW for fixing SB. */
	if (sb_mode != RM_CHECK) {
		flags = host->flags;
		if (aal_device_reopen(host, host->blksize, O_RDWR)) {
			aal_fatal("Failed to reopen file \"%s\" with O_RDRW: %s",
				  host->name, strerror(errno));
			return -EIO;
		}
		
		reop = 1;
	}

	repair->mode = sb_mode;
	if ((res = repair_fs_open(repair, host, host)))
		return res;

	if (repair->fs == NULL) {
		aal_fatal("Cannot open the FileSystem on (%s).", host->name);
		return res;
	}

	if (!reop) {
		/* Reopen device RW for replaying. */
		flags = host->flags;
		if (aal_device_reopen(host, host->blksize, O_RDWR)) {
			aal_fatal("Failed to reopen file \"%s\" with O_RDRW: %s",
				  host->name, strerror(errno));
			res = -EIO;
			goto error_close_fs;
		}
	}
	
	if ((res = repair_fs_replay(repair->fs, &repair->num_replayed_tx)) < 0)
		goto error_close_fs;
	
	repair_error_count(repair, res);
	
	/* Leave device RW if not CHECK mode. */
	reiser4_journal_sync(repair->fs->journal);

	if (reiser4_fs_sync(repair->fs))
		return -EIO;
	
	if (fs_mode == RM_CHECK) {
		if (aal_device_reopen(host, host->blksize, flags)) {
			aal_fatal("Failed to reopen file \"%s\" with 0%o: %s",
				  host->name, flags, strerror(errno));
			res = -EIO;
			goto error_close_fs;
		}
	}
	
	repair->sb_fixable = repair->fixable;
	repair->fixable = 0;
	repair->mode = fs_mode;

	repair_error_count(repair, res);
	
	repair->fs->tree->mpc_func = misc_mpressure_detect;
	
	aal_stream_init(&stream, NULL, &memory_stream);
	
	repair_master_print(repair->fs->master, &stream, 
			    misc_uuid_unparse);
	
	aal_stream_format(&stream, "\n");
	
	repair_format_print(repair->fs->format, &stream);
	aal_stream_format(&stream, "\n");
	
	aal_mess("Reiser4 fs was detected on %s.\n%s",
		repair->fs->device->name, (char *)stream.entity);
	
	aal_stream_fini(&stream);
	
	/* Initialize the backup. */
	len = reiser4_format_get_len(repair->fs->format);
	if ((res = backup_init(backup, host, len)))
		goto error_close_fs;

	return 0;
	
 error_close_fs:
	reiser4_fs_close(repair->fs);
	repair->fs = NULL;
	return res;
}

static errno_t fsck_check_fini(repair_data_t *repair, 
			       uint8_t sb_mode, 
			       uint8_t fs_mode, 
			       errno_t result)
{
	reiser4_status_t *status;
	aal_device_t *device;
	uint64_t state;
	int flags;

	aal_assert("vpf-1338", repair != NULL);
	aal_assert("vpf-1339", repair->fs != NULL);
	aal_assert("vpf-1340", repair->fs->status != NULL);
	
	device = repair->fs->device;
	flags = device->flags;
	
	if (aal_device_reopen(device, device->blksize, O_RDWR))
		return -EIO;
	
	/* Fix the status block. */
	status = repair->fs->status;

	if (repair->fatal) {
		state = FS_DAMAGED;
	} else if (repair->fixable || repair->sb_fixable) {
		state = FS_CORRUPTED;
	} else {
		state = FS_OK;
	}
	repair_status_update(status, state);

	if (state != FS_OK || repair->num_replayed_tx != 0)
		/* fsck doesn't work with mirrors */
		repair_status_extended_update(status, FSE_MIRRORS_NOT_SYNCED);

	if (reiser4_status_sync(repair->fs->status))
		return -EIO;

	if (!result && (sb_mode != RM_CHECK || fs_mode != RM_CHECK)) {
		/* If there was no backup openned or some fields have 
		   been changed, reopen the backup. */
		if (!(repair->fs->backup = repair_backup_reopen(repair->fs))) {
			aal_fatal("Failed to reopen backup.");
			return -EIO;
		}

		if (reiser4_backup_sync(repair->fs->backup))
			return -EIO;
	}

	if (aal_device_reopen(device, device->blksize, flags))
		return -EIO;

	return 0;
}

int main(int argc, char *argv[]) {
	struct aal_device_ops fsck_ops = file_ops;
	aal_device_t *device;
	repair_data_t repair;
	
	fsck_parse_t parse_data;
	errno_t ex = NO_ERROR;
	int stage = 0;
	errno_t res = 0;

	memset(&parse_data, 0, sizeof(parse_data));
	memset(&repair, 0, sizeof(repair));
	
	if ((ex = fsck_init(&parse_data, &fsck_ops, argc, argv)) != NO_ERROR)
		exit(ex);

	device = parse_data.host_device;
	
	if (parse_data.fs_mode == RM_BACK){
		if (aal_device_reopen(device, device->blksize, O_RDWR)) {
			ex = OPER_ERROR;
			goto free_device;
		}

		if ((res = backup_rollback(parse_data.backup, device)))
			ex = OPER_ERROR;

		goto free_device;
	}
	
	fsck_time("fsck.reiser4 started at");
	
	/* SB_mode is specified, otherwise  */
	repair.flags = 
		fsck_opt(&parse_data, FSCK_OPT_DEBUG) << REPAIR_DEBUG |
		fsck_opt(&parse_data, FSCK_OPT_WHOLE) << REPAIR_WHOLE |
		fsck_opt(&parse_data, FSCK_OPT_OLD) << REPAIR_WHOLE |
		fsck_opt(&parse_data, FSCK_OPT_OLD) << REPAIR_NO_MKID |
		fsck_opt(&parse_data, FSCK_OPT_NOMKID) << REPAIR_NO_MKID |
		fsck_opt(&parse_data, FSCK_OPT_YES) << REPAIR_YES;
		
	repair.bitmap_file = parse_data.bitmap_file;
	
	res = fsck_check_init(&repair, device, parse_data.backup, 
			      parse_data.sb_mode, parse_data.fs_mode);
	
	if (res || repair.fatal)
		goto free_libreiser4;
		
	stage = 1;
	
	res = repair_check(&repair);

	/* Even if there was some problems on fs check, fini must be done. */
	res |= fsck_check_fini(&repair, parse_data.sb_mode, 
			       parse_data.fs_mode, res);
	
	fsck_time("fsck.reiser4 finished at");
    
	fprintf(stderr, "Closing fs...");
	reiser4_fs_close(repair.fs);
	repair.fs = NULL;
	fprintf(stderr, "done\n");
    
 free_libreiser4:
	libreiser4_fini();
    
 free_device:
	if (device) {
		if (aal_device_sync(device)) {
			aal_fatal("Cannot synchronize file \"%s\"", 
				  device->name);
			ex = OPER_ERROR;
		}
		aal_device_close(device);
	}
	
	fprintf(stderr, "\n");
	
	/* Report about the results. */
	if (res < 0 || ex == OPER_ERROR) {
		ex = OPER_ERROR;
		fprintf(stderr, "Operational error occurred while fscking.\n");
		goto free_fsck;
	} 
	
	if (parse_data.fs_mode == RM_BACK)
		goto free_fsck;
	
	if (repair.sb_fixable) {
		/* No fatal corruptions in SB, but some fixable ones. */
		fprintf(stderr, "%llu fixable corruptions were detected in "
			"the SuperBlock. Run with --fix option to fix them.\n",
			repair.sb_fixable);
		
		ex = FIXABLE_ERROR;
	}

	if (repair.fatal) {
		/* Some fatal corruptions in disk format or filesystem. */
		if (parse_data.fs_mode == RM_BUILD && stage) {
			/* Only if no metadata are found. */
			fprintf(stderr, "NO REISER4 METADATA ARE FOUND. "
				"FS RECOVERY IS NOT POSSIBLE.\n");
		} else {
			fprintf(stderr, "%llu fatal corruptions were detected in %s. "
				"Run with %s option to fix them.\n", repair.fatal, 
				stage ? "FileSystem" : "SuperBlock", 
				stage ? "--build-fs" : "--build-sb");
		}
		ex = stage ? FATAL_ERROR : FATAL_SB_ERROR;
	} else if (repair.fixable) {
		/* Some fixable corruptions in filesystem. */
		fprintf(stderr, "%llu fixable corruptions were detected in "
			"the FileSystem. Run with --fix option to fix them.\n",
			repair.fixable);
		
		ex = FIXABLE_ERROR;
	} else if (!repair.sb_fixable) {
		if (parse_data.fs_mode != RM_CHECK && 
		    fsck_opt(&parse_data, FSCK_OPT_RO))
		{
			fprintf(stderr, "\nThe filesystem was mounted RO. It is "
				"better to umount and mount it again.\n");
		}
		
		fprintf(stderr, "FS is consistent.\n");
	}
	
	fprintf(stderr, "\n");

 free_fsck:
	fsck_fini(&parse_data);
	
	return ex;
}
